---
title: "dataViz"
author: "Najeeb Khan, Hafik Arhan Kama??, Swaraj Oturkar, Elif Erbil"
date: "12/4/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_knit$set(root.dir = '~/Google Drive/TUM Session/Semester One/Data Visualisation in R/')

```

### Getting Data 
```{r echo=FALSE, include=FALSE}
library(data.table)
library(magrittr)
library(ggplot2)
library(reshape)
<<<<<<< HEAD
library(readxl)
=======
library(tidyr)
>>>>>>> 7b041f9bac6123d96f3a9cf2be6e814225fced3d
```
```{r echo=TRUE}
genotype_file <- 'Data/eqtl/genotype.txt'
growth_file <- 'Data/eqtl/growth.txt'
genotype <- as.data.table(read.delim(genotype_file))
growth <- as.data.table(read.delim(growth_file))
```
### Checking the integrity of the data in both of the tables

```{r echo=TRUE}
## Genotype Data
names <- c("strain")
for(i in 1:1000)
{
  names <- c(names, paste("mrk", i, sep = "_"))
}
colnames(genotype) <- names

### Calculating the proportion of Wild Isolate and Lab Strain
genotype[, labProp := apply(genotype, MARGIN = 1, FUN = function(row) mean(grepl(unname(unlist(row)), pattern = 'Lab strain')))]

genotype[, wildProp := apply(genotype, MARGIN = 1, FUN = function(row) mean(grepl(unname(unlist(row)), pattern = 'Wild isolate')))]

## Growth Data

### Checking the column class
sapply(growth, class)

### Melting the data to make it tidy
growth_melt <- melt(growth, id.vars = 'strain', variable.name = 'env', value.name = 'rate')

### Plotting boxplots to recognize any outliers
ggplot(growth_melt, aes(env, rate)) + geom_boxplot() + theme_minimal() +
  labs(title = "Visualisation of Growth Data") +
  xlab("Environment") +
  ylab("Rate of Growth")

```

### Getting insight into how proportion of each parent strain effects the growth
We investigate how the proportion of parent strain (from Wild Isolate and Lab Strain) is responsible for effecting the growth in a particular medium.

```{r echo=TRUE}
## Merging the two datasets on the basis of segregants
growth_geno <- merge(genotype[, c(1, 1002, 1003)], growth)

## Melting the merged table
growth_geno_melt <- melt(growth_geno, id.vars = c('strain', 'labProp', 'wildProp'), variable.name = 'env', value.name = 'rate')

## PLotting the data points as barplot distributed over environments
ggplot(growth_geno_melt, aes(x = wildProp, y=rate)) + geom_point() + theme_minimal() +
  facet_wrap(~env)

ggplot(growth_geno_melt, aes(x = labProp, y=rate)) + geom_point() + theme_minimal() +
  facet_wrap(~env)

<<<<<<< HEAD
## PLotting the data points as a single scatterplot
ggplot(growth_geno_melt, aes(x = wildProp, y = rate, color=env)) + geom_point() + theme_minimal()

## Growth and Marker plots

### Merging growth and genotype data with markers
growth_marker <- merge(genotype, growth)
growth_marker_melt <- melt(growth_marker, id.vars = colnames(growth_marker)[1:1003], variable.name = "env", value.name="rate")

### Melting the markers in one column to make data tidy
growth_marker_melt <- melt(growth_marker_melt, id.vars = c("strain", "env", "rate", "labProp", "wildProp"), variable.name = "marker_number", value.name = "parent_strain")

### Cleaning the marker number column
growth_marker_melt[, marker_number := as.integer(gsub("mrk_", "", marker_number))]

### Plotting marker number with the mean of the rate in each of those markers
growth_marker_melt[, mean_growth_per_marker := mean(rate, na.rm = T), by=marker_number]
=======
```
>>>>>>> 7b041f9bac6123d96f3a9cf2be6e814225fced3d

### Getting insight into how genes are expressed on different environments
We investigate how the expression rate of each gene and gene type varies in different environments. 

<<<<<<< HEAD
### They will be a straightline because for each segregant there is a single growth => for all 1000 markers there is only a single growth 

```

### Getting insight into how the gene expression is affected by the environment

```{r echo=True}
### Getting the gene expression  and gene data as they are connected
gene_expression_file <- 'Data/eqtl/expression.txt'
gene_expression <- as.data.table(read.table(gene_expression_file), keep.rownames = T)
colnames(gene_expression) <- as.character(gene_expression[1, ])
unname(gene_expression)

gene_file <- 'Data/eqtl/gene.txt'
gene_data <- read.csv(gene_file, sep = '\t')

### Merging expression and gene data
gene_expression_merge <- merge(gene_expression, gene_data)

=======
```{r echo=TRUE}
gene_file <- 'Data/eqtl/gene.txt'
expression_file <- 'Data/eqtl/expression.txt'
gene <- as.data.table(read.delim(gene_file))
expression <- as.data.table(read.delim(expression_file, comment.char = "#"))
summary(gene)
summary(expression)
expression[, gene := gene$name]
expression[, gene_type := gene$type]
summary(expression)
head(gene, n = 20)

head(expression, n = 20)
head(genotype$strain, n = 20)

expression_tidy <- gather(expression, medium_strand, expression_rate, c("YPD.seg_01B":"YPMalt.seg_45C"))
expression_tidy <- as.data.table(expression_tidy)
expression_tidy[, medium_strand := .(gsub("_", "", medium_strand))]
head(expression_tidy, n = 10)
expression_tidy <- as.data.table(separate(as.data.frame(expression_tidy), medium_strand, into = c("medium", "strand")))
expression_tidy

gene_expr <- unique(expression_tidy[, mean_expr := .SD[, mean(expression_rate)], by = c("gene","medium")])

#expression rates in each medium
YDP_expr <- unique(expression_tidy[medium == "YPD", .SD[,.(medium, mean(expression_rate))], by = gene])
YDPBPS_expr <-unique(expression_tidy[medium == "YPDBPS", .SD[,.(medium, mean(expression_rate))], by = gene])
YDPRapa_expr <- unique(expression_tidy[medium == "YPDRapa", .SD[,.(medium, mean(expression_rate))], by = gene])
YPE_expr <- unique(expression_tidy[medium == "YPE", .SD[,.(medium, mean(expression_rate))], by = gene])
YPMalt_expr <- unique(expression_tidy[medium == "YPMalt", .SD[,.(medium, mean(expression_rate))], by = gene])

ggplot(YDP_expr, aes(gene, V2)) + geom_point()
ggplot(YDPBPS_expr, aes(gene, V2)) + geom_point()
ggplot(YDPRapa_expr, aes(gene, V2)) + geom_point()
ggplot(YPE_expr, aes(gene, V2)) + geom_point()
ggplot(YPMalt_expr, aes(gene, V2)) + geom_point()

#average expression of each gene in each medium in one graph
mean_expr_dt <- unique(gene_expr[,.(medium, mean_expr, gene_type), by=gene])
ggplot(mean_expr_dt[1:50], aes(gene, mean_expr)) + geom_point(aes(color = medium))
ggplot(mean_expr_dt[1:50], aes(gene, 2^mean_expr)) + geom_point(aes(color = medium))

#Average expression of each gene type in each environment
ggplot(mean_expr_dt, aes(gene_type,2^mean_expr)) + geom_boxplot() + facet_wrap(~medium)
>>>>>>> 7b041f9bac6123d96f3a9cf2be6e814225fced3d
```