---
title: "Case Study 1"
author: "Najeeb Khan, Hafik Arhan Kamac, Swaraj Oturkar, Elif Erbil"
date: "12/4/2018"
output:
  pdf_document: default
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/Google Drive/TUM Session/Semester One/Data Visualisation in R/')
```
```{r echo=FALSE, include=FALSE}
library(data.table)
library(magrittr)
library(ggplot2)
library(reshape)
library(readxl)
library(tidyr)
```
#Introduction
In this case study we have examined the relations between a genome, gene, expression rate and growth rate in a set of yeast segregants and environments. We used the data given to us to answer questions such as the relation between genotype and environment on the growth rate of the yeast, which can also be related to the expression of the genes which result with translation of them and production of proteins. In order to analyze our data, we first tidied each table so that each column has a variable and each row has an observation. We will be presenting the methods we used to tidy the data and the analysis we made.

#Tidying Data
```{r echo=TRUE}
gene_file <- 'Data/eqtl/gene.txt'
expression_file <- 'Data/eqtl/expression.txt'
gene <- as.data.table(read.delim(gene_file))
expression <- as.data.table(read.delim(expression_file, comment.char = "#"))

name <- gene[,.(name)]
type <- gene[,.(type)]
expression[, gene := name][, gene_type := type]

#tidy expression.txt file
expression_tidy <- gather(expression, medium_strand, expression_rate, c("YPD.seg_01B":"YPMalt.seg_45C"))
expression_tidy <- as.data.table(expression_tidy)
expression_tidy[, medium_strand := .(gsub("_", "", medium_strand))]
head(expression_tidy, n = 10)
expression_tidy <- as.data.table(separate(as.data.frame(expression_tidy), medium_strand, into = c("medium", "strand")))

#tidied version of expression data
head(expression_tidy)
```
### Getting Data 

```{r echo=TRUE}
#TODO: Can you tidy the code here please?
genotype_file <-'Data/eqtl/genotype.txt'
growth_file <- 'Data/eqtl/growth.txt'
gene_file <- 'Data/eqtl/gene.txt'
marker_file <- 'Data/eqtl/marker.txt'

genotype <- as.data.table(read.delim(genotype_file))
growth <- as.data.table(read.delim(growth_file))
gene <- as.data.table(read.delim(gene_file))
marker <- as.data.table(read.delim(marker_file))

```

### Checking the integrity of the data in both of the tables

```{r echo=TRUE}
## Genotype Data
#### names <- c("strain")
#### for(i in 1:1000)
#### {
####  names <- c(names, paste("mrk", i, sep = "_"))
#### }
#### colnames(genotype) <- names

### Calculating the proportion of Wild Isolate and Lab Strain
genotype[, labProp := apply(genotype, MARGIN = 1, FUN = function(row) mean(grepl(unname(unlist(row)), pattern = 'Lab strain')))]

genotype[, wildProp := apply(genotype, MARGIN = 1, FUN = function(row) mean(grepl(unname(unlist(row)), pattern = 'Wild isolate')))]

### Checking the column class
sapply(growth, class)

### Melting the data to make it tidy
growth_melt <- melt(growth, id.vars = 'strain', variable.name = 'env', value.name = 'rate')
```


#Analysis

### Effect of Environment on Growth
We investigate how the environment affect growth rate of the segregants.
```{r echo=TRUE}
# Plotting boxplots to recognize any outliers
ggplot(growth_melt, aes(env, rate)) + geom_boxplot() + theme_minimal() +
  labs(title = "Visualisation of Growth Data") +
  xlab("Environment") +
  ylab("Rate of Growth")

```
In the generated box-plot we see that the distribution of the growth rates varies in every environment. Since we have the growth data of the same strain in different environments, the box-plot shows that growth rate is affected by the environment and most of the strains were able to grow more in the environment YPD.

### Getting insight into how proportion of each parent strain affects the growth
In this section we investigate how the proportion of parent strain (from Wild Isolate and Lab Strain) is responsible for effecting the growth in a particular medium.

```{r echo=TRUE}
## Merging the two datasets on the basis of segregants
growth_geno <- merge(genotype[, c(1, 1002, 1003)], growth)

gene_marker <- merge(marker,gene,by = "chrom",allow.cartesian = T)
setnames(gene_marker,"start.x", "marker_start")
setnames(gene_marker,"start.y", "gene_start")
setnames(gene_marker,"end.x", "marker_end")
setnames(gene_marker,"end.y", "gene_end")

## Melting the merged table
growth_geno_melt <- melt(growth_geno, id.vars = c('strain', 'labProp', 'wildProp'), variable.name = 'env', value.name = 'rate')

##Finding those markers which are on gene either completely or partially
gene_marker_filtered <- subset(gene_marker,!(marker_start > gene_end | marker_end < gene_start))

## TODO: Find the type of strain on each of the genes that are expressed through the marker
colnames(genotype)

genotype_melt <- melt(genotype, id.vars="strain", variable.name = "id", value.name = "parent_strain")

new_table <- merge(x = genotype_melt, y = gene_marker_filtered, by = "id", allow.cartesian = T)
subset_new_table <- new_table[, .(id, strain, parent_strain)]
subset_new_table <- unique(subset_new_table)

casted <- dcast(subset_new_table, ... ~ id, value.var =  "parent_strain")

casted[, wildProp := apply(casted, MARGIN = 1, FUN = function(row) mean(grepl(unname(unlist(row)), pattern = 'Wild isolate')))]

#### Merging both of the databases
casted_merged <- merge(growth, casted[, .(strain, wildProp)], by = "strain")
### Melting the data
casted_merged <- melt(casted_merged, id.vars = c('strain', 'wildProp'), variable.name = 'env', value.name = 'rate')


ggplot(casted_merged, aes(x = wildProp, y=rate)) + geom_point() + theme_minimal() +
  facet_wrap(~env)

## PLotting the data points as barplot distributed over environments
ggplot(growth_geno_melt, aes(x = wildProp, y=rate)) + geom_point() + theme_minimal() +
  facet_wrap(~env)

ggplot(growth_geno_melt, aes(x = labProp, y=rate)) + geom_point() + theme_minimal() +
  facet_wrap(~env)

## PLotting the data points as a single scatterplot
ggplot(growth_geno_melt, aes(x = wildProp, y = rate, color=env)) + geom_point() + theme_minimal()

## Growth and Marker plots

### Merging growth and genotype data with markers
growth_marker <- merge(genotype, growth)
growth_marker_melt <- melt(growth_marker, id.vars = colnames(growth_marker)[1:1003], variable.name = "env", value.name="rate")

### Melting the markers in one column to make data tidy
growth_marker_melt <- melt(growth_marker_melt, id.vars = c("strain", "env", "rate", "labProp", "wildProp"), variable.name = "marker_number", value.name = "parent_strain")

### Cleaning the marker number column
growth_marker_melt[, marker_number := as.integer(gsub("mrk_", "", marker_number))]

### Plotting marker number with the mean of the rate in each of those markers
growth_marker_melt[, mean_growth_per_marker := mean(rate, na.rm = T), by=marker_number]
```

We have calculated the percentage of lab and wild strains of each segregant to observe if a segregant with a majority of a certain genotype has a higher growth rate in a specific environment. Our findings showed that in none of the environments, we couldn???t observe a relation between increasing genotype percentage and growth rate. We were expecting to get such a result since we know each gene has a different expression rate and just by knowing the proportion of a genotype, we cannot be able to conclude the effect of it on the growth rate. Therefore, in the next part we will be examining relation between gene expression and growth rate with respect to whether the genotype of a each strain effects its expression rate.

### Getting insight into how proportion of each parent strain affects the growth
In this section we determined the markers which represent one or more genes.

```{r echo=TRUE}
#TODO: Swaraj's code that matches markes and genes goes HERE
```
We have determined the markers which contains one or more gene by checking the indices of markers and genes and the ones that overlapped has the genes. Therefore we were able to determine the genotype of a gene and we were able to use these in our analysis of growth and expression rates.

### Investigating the genotype affect on the growth rate of a segregant
In this section we used the markers that corespond to a gene to check if the genotype of the genes affect the growth rate
```{r echo=TRUE}
#TODO: Najeeb's code that calculates differences for each marker for comparing genotypes
```

### Determining the dependency of gene expression on environments 
We investigate how the expression rate of each gene and gene type varies in different environments. 

### They will be a straightline because for each segregant there is a single growth => for all 1000 markers there is only a single growth 

### Getting insight into how the gene expression is affected by the environment

```{r echo=True}
### Getting the gene expression  and gene data as they are connected
gene_expression_file <- 'Data/eqtl/expression.txt'
gene_expression <- as.data.table(read.table(gene_expression_file), keep.rownames = T)
colnames(gene_expression) <- as.character(gene_expression[1, ])
unname(gene_expression)

gene_file <- 'Data/eqtl/gene.txt'
gene_data <- read.csv(gene_file, sep = '\t')

### Merging expression and gene data
gene_expression_merge <- merge(gene_expression, gene_data)
```
```{r echo=TRUE}

#mean expression rate of each gene in each environments 
mean_expr_gene <- unique(expression_tidy[, .(gene_type, mean_expr = .SD[, mean(expression_rate)]), by = c("gene","medium")])

<<<<<<< HEAD
expression_tidy <- gather(expression, medium_strand, expression_rate, c("YPD.seg_01B":"YPMalt.seg_45C"))
expression_tidy <- as.data.table(expression_tidy)
expression_tidy[, medium_strand := .(gsub("_", "", medium_strand))]
head(expression_tidy, n = 10)
expression_tidy <- as.data.table(separate(as.data.frame(expression_tidy), medium_strand, into = c("medium", "strand")))
expression_tidy

gene_expr <- unique(expression_tidy[, mean_expr := .SD[, mean(expression_rate)], by = c("gene","medium")])

#expression rates in each medium
YDP_expr <- unique(expression_tidy[medium == "YPD", .SD[,.(medium, mean(expression_rate))], by = gene])
YDPBPS_expr <-unique(expression_tidy[medium == "YPDBPS", .SD[,.(medium, mean(expression_rate))], by = gene])
YDPRapa_expr <- unique(expression_tidy[medium == "YPDRapa", .SD[,.(medium, mean(expression_rate))], by = gene])
YPE_expr <- unique(expression_tidy[medium == "YPE", .SD[,.(medium, mean(expression_rate))], by = gene])
YPMalt_expr <- unique(expression_tidy[medium == "YPMalt", .SD[,.(medium, mean(expression_rate))], by = gene])

ggplot(YDP_expr, aes(gene, V2)) + geom_point()
ggplot(YDPBPS_expr, aes(gene, V2)) + geom_point()
ggplot(YDPRapa_expr, aes(gene, V2)) + geom_point()
ggplot(YPE_expr, aes(gene, V2)) + geom_point()
ggplot(YPMalt_expr, aes(gene, V2)) + geom_point()

#average expression of each gene in each medium in one graph
mean_expr_dt <- unique(gene_expr[,.(medium, mean_expr, gene_type), by=gene])
ggplot(mean_expr_dt[1:50], aes(gene, mean_expr)) + geom_point(aes(color = medium))
ggplot(mean_expr_dt[1:50], aes(gene, 2^mean_expr)) + geom_point(aes(color = medium))

#Average expression of each gene type in each environment
ggplot(mean_expr_dt, aes(gene_type,2^mean_expr)) + geom_boxplot() + facet_wrap(~medium)
```


## Warm Up Exercise
``` {r echo=TRUE}
marker <- read.delim("Data/eqtl/marker.txt")
growth <- read.delim("Data/eqtl/growth.txt")
genotype <- read.delim("Data/eqtl/genotype.txt")

getMeasure <- function(strt, chr, genotype, growth, df)
{
  cols <- which(df$chrom == chr & df$start == strt)
  
  if(cols == 1)
    cols = cols + 1
  mygeno <- genotype[, cols]
  
  # Creating temporary database for Lab strain
  temp <- as.data.table(melt(growth[mygeno == "Lab strain", ], id.vars = "strain"))
  med_lab <- temp[, median(value, na.rm = T), by=variable]
  
  # Creating temporary database for Wild Isolate
  temp <- as.data.table(melt(growth[mygeno == "Wild isolate", ], id.vars = "strain"))
  med_wild <- temp[, median(value, na.rm = T), by=variable]
  
  return(med_wild[, V1] - med_lab[, V1])
  
}

num_markers <- 1000
marker_dt <- as.data.table(marker)
difference <- list()

filtered_markers <- gene_marker_filtered[, .(chrom, id, marker_start)]
filtered_markers <- unique(filtered_markers)
num_filtered_markers <-  dim(filtered_markers)[1]
print(num_filtered_markers)

for(i in 1:num_filtered_markers){
  
  c <- as.character(filtered_markers[i, 1]$chrom)
  s <- as.integer(filtered_markers[i, 3]$marker_start)
  x <- getMeasure(s, c, genotype, growth, marker)
  dim(x) <- c(1,5)
  x <- as.data.table(x)
  
  difference <- rbindlist(list(difference, x))
}


difference

colnames(difference) <- colnames(growth)[2:6]
difference[, mrk_no := c(1:887)]
ggplot(difference, aes(mrk_no, YPD)) + geom_point() + theme_minimal()

plot(YPMalt ~ mygeno[strain], data=growth)
```
=======
diff_expr_gene <- mean_expr_gene[, .(mean_expr, diff_expr = (max(mean_expr) - min(mean_expr)), medium), by=gene]
ordered_genes <- setorder(diff_expr_gene, diff_expr)[, .(gene, diff_expr, medium)]
ordered_genes <- unique(ordered_genes[, diff_expr, by = gene])
#the genes that are most environment independent
head(ordered_genes, n=3)[,gene]
#the genes that are most environment dependent
tail(ordered_genes, n=3)[,gene]
```

We have calculated differences between each gene's mean expression rate for each environment. We assumed that the gene with the lowest difference is the one that is not dependent on the environment since the expression rate is nearly the same at each environment. So we have ordered all genes to find out which genes are less dependent on the environment. We found out that the genes YML035C, YLR035C, YPR029C are the ones that are least depent ones and the genes YNL117W, YMR107W, YJR005C-A are the most dependant ones on the environment.

### Understanding the affect of the type of the gene on the expression rate
In this section we examined how the different types of genes defer in expression rate in different environments.

```{r echo=TRUE}
#Expression rate of each type of genes in each environment
gene_type_data <- unique(mean_expr_gene[,.(medium, mean_expr, gene_type), by=gene])
ggplot(gene_type_data, aes(gene_type, mean_expr)) + geom_boxplot() + facet_wrap(~medium) 
```

We have seen that regardless of the environment, genes with the ORF-T type are the highest expressed genes. This is partially due to the type of the gene, since ORF type genes are used in translation, it is only logical that they are the most expressed genes.

### Correlation between genotype and expression rate of a gene
We have tried to find if a correlation exists between the expression rate of a gene and if the gene comes from a wild or lab isolate.

```{r echo=TRUE}
#TODO: Change Najeeb's code to calculate differences for each marker for comparing genotypes with expression rate this time
```

#Conclusion
